### PR TITLE
Allow for leading slashes in .distignore, and deal gracefully with symlinks

### LINKED ISSUE #57: Symlink in ignored directory causing out of memory error
## Bug Report

- [x] Yes, I reviewed the [contribution guidelines](https://make.wordpress.org/cli/handbook/contributing/).
- [x] Yes, more specifically, I reviewed the guidelines on [how to write clear bug reports](https://make.wordpress.org/cli/handbook/bug-reports/).

**Describe the current, buggy behavior**

I have a project structure with a recursive symlink:

```
project-dir/my-plugin.php
project-dir/wp-content/plugins/my-plugin/
```

where `wp-content/plugins/my-plugin/` is a symlink to `project-dir/`.

Previously I was using `dist-archive` on the `src` directory and now I'm trying to use it on the entire project directory, with appropriate an `.distignore` in the project root. [PhpStorm project screenshot](https://brianhenryie.s3.amazonaws.com/2021/wp-plugin-project-structure.png)

My `.distignore` file contains `wp-content` in it.

Still, running `wp dist-archive . out.zip` causes a `PHP Fatal error`:

```
PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 121634848 bytes) in /path/to/project-dir/vendor/wp-cli/wp-cli/php/WP_CLI/Process.php on line 75
```

When I delete the symlink, I am able to use `dist-archive` without issue.

**Describe what you would expect as the correct outcome**

I expect the `wp-content` folder to have no bearing on the operation because it is listed in the `.distignore` file.

**Let us know what environment you are running this on**

MacOS M1,  Monterey (12.3.1), PHP 8.0 via Homebrew, Zsh shell.

```
OS:     Darwin 21.4.0 Darwin Kernel Version 21.4.0: Fri Mar 18 00:47:26 PDT 2022; root:xnu-8020.101.4~15/RELEASE_ARM64_T8101 arm64
Shell:  /bin/zsh
PHP binary:     /opt/homebrew/bin/php
PHP version:    8.0.19
php.ini used:   /opt/homebrew/etc/php/8.0/php.ini
MySQL binary:   /opt/homebrew/bin/mysql
MySQL version:  mysql  Ver 15.1 Distrib 10.7.3-MariaDB, for osx10.17 (arm64) using  EditLine wrapper
SQL modes:      STRICT_TRANS_TABLES,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION
WP-CLI root dir:        /Users/brianhenry/Sites/bh-wc-disable-free-gifts-per-product/vendor/wp-cli/wp-cli
WP-CLI vendor dir:      /Users/brianhenry/Sites/bh-wc-disable-free-gifts-per-product/vendor
WP_CLI phar path:       
WP-CLI packages dir:    /Users/brianhenry/.wp-cli/packages/
WP-CLI global config:   
WP-CLI project config:  /Users/brianhenry/Sites/bh-wc-disable-free-gifts-per-product/wp-cli.yml
WP-CLI version: 2.6.0

```

**Provide a possible solution**

Using the `zip` flag:

```
-y   store symbolic links as the link instead of the referenced file
```

Change:

`$cmd = "zip -r '{$archive_file}' {$archive_base} {$excludes}";`

to:

`$cmd = "zip -y -r '{$archive_file}' {$archive_base} {$excludes}";`

This works.

But I presume it would prevent the `dist-ignore` command from working with plugins that use symlinks internally to the plugin (i.e. not just dev tools and environment) during development.

Alternatively, I could copy all files to system temp before archiving them.

### PR DESCRIPTION
**Pull request by BrianHenryIE**
This is tests and fixes for PRs #58 and #59.

The tests for 58 were relevant to those needed for 59.
The code in 59 fixed some of the issue in 58.

Closes #57 and #47.

PHPUnit tests are added to this repo for the first time and should be helpful for work that might be done on #15 and #44.

### CODE REVIEW

**Comment by BrianHenryIE** [comment]
Dammit, tests working locally on MacOS and not on GitHub Actions! 

Syntax differences between `tar` on different systems?

File: features/dist-archive.feature [hunk 1]
```diff
@@ -203,6 +203,114 @@ Feature: Generate a distribution archive of a project
       | zip     | zip       | unzip     |
       | targz   | tar.gz    | tar -zxvf |
 
+  Scenario Outline: Ignores files specified with absolute path and not similarly named files
```
Comment (wojsmol) [inline features/dist-archive.feature:206]: In other test we  are using `Scenario:` instead of `Scenario Outline:`

Comment (BrianHenryIE) [reply features/dist-archive.feature:206]: `Scenario Outline:` is used where there is an `Examples:` table at the end of the test, like a PHPUnit `@dataProvider`. I'm pretty sure it's correct here.

https://docs.behat.org/en/latest/user_guide/writing_scenarios.html#scenario-outlines

**Comment by BrianHenryIE** [comment]
I needed to use `php_uname( 's' )` to test for Linux because `tar`'s syntax and flags differ between MacOS and Linux.

`tar`'s `--anchored` is not available on MacOS.
On MacOS `tar`, to anchor, exclusions are written beginning with `^` regex character.

**Comment by danielbachhuber** [comment]
Thanks @BrianHenryIE, appreciate your work on this.

I'll review this at depth when I have some time in the near future.

**Review by danielbachhuber — APPROVED** [review]
This looks great me, @BrianHenryIE.

Thanks again for your effort on it, and your thorough tests. The tests pass for me locally on Mac.

File: features/dist-archive.feature (hunk 1 above)
Comment (danielbachhuber) [reply features/dist-archive.feature:206]: Yep, `Scenario Outline:` lets you run multiple variables against the same scenario.

**Pull request merged**